btk/create_observing_generator.py

```diff
@@ -1,109 +1,43 @@
-        shape (tuple): shape of the image
-        center_pix (tuple): position of the reference pixel used as the center of the
-                            affine transform for the wcs.
-        center_sky (float):
-        naxis (int):
-        projection(str):
-
-    Returns:
-        wcs: WCS
-    """
-    if center_pix is None:
-        center_pix = [(s + 1) / 2 for s in shape]
-    if center_sky is None:
-        center_sky = [0 for _ in range(naxis)]
-    if projection is None:
-        projection = "TAN"
-    w = WCS.WCS(naxis=2)
-    w.wcs.ctype = ["RA---" + projection, "DEC--" + projection]
-    w.wcs.crpix = center_pix
-    w.wcs.cdelt = [pixel_scale for _ in range(naxis)]
-    w.wcs.crval = center_sky
-    w.array_shape = shape
-    return w
-
-
-def default_obs_conditions(Args, band):
-    """Returns the default observing conditions from the WLD package
-    for a given survey_name and band.
-
-    Args:
-        Args: A `btk.config.SimulationParams` object containing parameters to generate blends
-        band: filter name to get observing conditions for.
-    Returns:
-        `survey`: Dictionary containing the observing conditions and WCS information.
-    """
-    survey = descwl.survey.Survey.get_defaults(
-        survey_name=Args.survey_name, filter_band=band
-    )
-    survey["center_sky"] = None
-    survey["center_pix"] = None
-    survey["projection"] = "TAN"
-    return survey
-
-
-def generate(Args, obs_function=None):
-    """Generates class with observing conditions in each band.
-
-    Args:
-        Args: Class containing input parameters.
-        obs_function: Function that outputs dict of observing conditions. If
-            not provided then the default `descwl.survey.Survey` values for the
-            corresponding Args.survey_name are used to create the
-            observing_generator.
-
-    Yields:
-        Generator with `btk.survey.Survey` class for each band.
-    """
-    while True:
+from btk.obs_conditions import DefaultObsConditions
+
+
+class ObservingGenerator:
+    def __init__(
+        self,
+        survey_name,
+        stamp_size,
+        obs_conditions=None,
+        verbose=False,
+    ):
+        """Generates class with observing conditions in each band.
+
+        Args:
+             survey_name (str): Name of the survey which should be available in descwl
+             obs_conditions: Class (not object) that returns observing conditions for
+                             a given survey and band. If not provided, then the default
+                             `descwl.survey.Survey` values for the corresponding
+                             survey_name are used to create the observing_generator.
+             stamp_size: In arcseconds.
+        """
+        if survey_name not in btk.survey.surveys:
+            raise KeyError("Survey not implemented.")
+
+        if obs_conditions == None:
+            obs_conditions = DefaultObsConditions(survey_name,"i",stamp_size)
```

Comment:
@thuiop , sorry I understand how this was confusing. My idea was to pass in the class (not object) of obs_conditions into the ObservingGenerator `__init__` and later in `__next__` to actually create the object and call it. But please do let me know what you think

I think this it will make it easier for multi resolution to not specify from the onset the "survey_name" and "band" like you do here.

Reply:
I think how it is right now is more consistent with how the sampling_function work for the blending generator. But I get what you mean and how I wrote it cannot be right as you need different conditions for each band ; you would need to save an obs_condition for each band. How about having a dictionnary containing the obs_conditions be created in the __init__ and then accessed in the __next__ to generate the survey ? That way, we may implement multi resolution as additionnal entries of this dictionnary ; there would be one observing_condition for each band in each survey.

Reply:
@thuiop thanks, I agree that it would be nice to keep consistency with how `sampling_function` works otherwise possibly causing confusion, so I struggled a bit with this change. 

Can you elaborate on your dictionary idea to make sure I'm understanding correctly? One thing is that I don't want to 'cache' the `btk.survey.Survey` from the beginning but allow it to change between calls of `__next__` (users might want different observing conditions between different batches if e.g. they are building algorithm to be robust between slightly different PSFs)

Reply:
You mean that users would be able to change the obs_conditions while keeping the same generator ? I think it is possible right now by accessing the fields of the WLDObsConditions object (like `obs_generator.obs_conditions.survey_name` for instance) ; at each next, the obs_conditions object is called, ie the get_survey function is called, meaning a whole new survey is generated from the parameters contained in the fields of the obs_conditions object.

As for the dictionnary, I mean that for each band you need a different obs_conditions (as you have a different psf and other parameters...). Those could be stored in some kind of dictionary which would contain the obs_conditions for all the bands (and possibly all the surveys with multi resolution).

Reply:
Take a look at my last commit ; I have coded what I described above (for the default case in particular).

Reply:
Thanks your solution should work well 👍